docs/prover/checking/sanity.md

```diff
@@ -85,218 +120,110 @@ rule tautology {
-since the tool could assume a starting state where the `root` is 0. This means
-that the invariant expression is not a tautology and the invariant passes. On
-the other hand the `aGE0` invariant, when run without the `--rule_sanity`
-option, will pass ([view report][aGE0-report]) but when we run it with the
-`--rule_sanity` `advanced` or `basic` options, the Prover creates a rule
-similar to the `aGE0Rule` below. This rule passes, indicating that the
-invariant expression is a tautology. The verification report shows that the
-invariant failed vacuity check ([view report][sanity-report]).
-
 ```{todo}
-Link to production, rather than staging.
+Does the tautology check also consider the bodies of contract functions?  For
+example, is `assert square(x) >= 0;` a tautology if `square` is a contract
+function that squares its input?
 ```
 
-[aGE0-report]: https://vaas-stg.certora.com/output/11775/871cf37193c75d27542b/?anonymousKey=dde443c4a806021716e863a454561a6ad1543d2e
-[sanity-report]: https://vaas-stg.certora.com/output/11775/4c4cb65f65c75f013c63/?anonymousKey=0b6a843857e6ead8e1bb1f11b984fb6e3e9fb6a8
-
-```solidity
-contract sanityCheck{
-    address root;
-    uint a;
-    constructor(address _root) {
-        require(_root != address(0));
-        root = _root;
-    }
-
-    function changeRoot (address root_) public {
-        require(root_ != address(0));
-        root = root_;
-    }
-}
+(sanity-trivial-invariant)=
+Trivial invariant checks
+------------------------
+
+The **Trivial invariant** sanity rule checks that invariants are not trivial.
+A trivial invariant is one that holds in all possible states, not just in
+reachable states.
+
+```{note}
+This check is currently called **assert-vacuity**
+```
+
+
+For example, the following invariant is trivial:
+
+```cvl
+invariant squaresNonNeg(int x)
+    x * x >= 0
 ```
 
+While it does hold in every reachable state, it also holds in every
+non-reachable state.  Therefore it could be more efficiently checked as a rule:
+
 ```cvl
+rule squaresNonNeg(int x) {
+    assert x * x >= 0;
+}
+```
+
+The rule version is more efficient because it can do a single check in an
+arbitrary state rather than separately checking states after arbitrary method
+invocations.
 
-  // Not a tautology
-  invariant rootNonZero()
-     root() != 0
+(sanity-assert-structure)=
+Assertion structure checks
+--------------------------
 
-  rule rootNonZero(){
-     assert root() != 0;
-  }
+The **assertion structure** sanity rule checks that complex assert statements
+can't be replaced with simpler ones.
 
-  // Tautology
-  invariant aGE0()
-     a() >= 0
+```{note}
+This check is currently called **assert-tautology**
+```
 
-  rule aGE0Rule{
-     assert a() >= 0;
-  }
+If an assertion expression is more complex than necessary, it can pass for
+misleading reasons.  For example, consider the following assertion:
 
+```cvl
+uint x;
+assert (x > 5) => (x >= 0);
```

Comment:
To add to that, the original assert-tautology was a misnomer.    A tautology is a formula that is always true, but if Q is only true in the current context and can be false if the same assert would be used in a different rule, the formula is not a tautology. 

Otherwise every assert in a correct rule is a tautology because it always evaluates to true.

Reply:
Okay, I got a bit confused as Trivial Invariant check is about invariants, not about asserts.   Does an `assert true;` or an `assert x>=0;` trigger any warnings at all?  Probably not. 

But with this rule `assert P => Q` triggers a warning if the code is unreachable in a state where P is false, even if P is not itself a tautology, right?   So something like `require x >= 3 ; assert x < 3 => y < 0;` triggers a warning, right?

Reply:
Re first comment: good point, I changed it to (x < 5) as you suggested.

Re second comment: that's why I changed the names, but what I've /now/ called assert tautology (formerly assert vacuity) does actually check for tautologies.

Re third comment: `assert true` and `assert x >= 0` would be caught by the assert tautology (formerly assert vacuity) check.

Your example `require x >= 3; assert x < 3 => y < 0;` would violate the "assertion structure" structure test because (x < 3) is false in context.  I will update the text to clarify that the checks are done in the context of the statement.